core/src/main/scala/kafka/server/ClientMetricsManager.scala

```diff
@@ -17,37 +37,73 @@ import scala.jdk.CollectionConverters._
 object ClientMetricsManager {
   private val _instance = new ClientMetricsManager
   def getInstance = _instance
+  def getCurrentTime = Calendar.getInstance.getTime.getTime
 
   def getSupportedCompressionTypes: List[java.lang.Byte] = {
     val compressionTypes = new ListBuffer[java.lang.Byte]
-    CompressionType.values.foreach(x => compressionTypes.append(x.id.toByte))
+    CompressionType.values.filter(x => x != CompressionType.NONE).foreach(x => compressionTypes.append(x.id.toByte))
     compressionTypes.toList
   }
 
+  // TODO: Needs to integrate with external plugin changes..
+  // if plugin is not configured, getMetricsSubscriptions and pushMetricSubscription needs to return errors
+  // to the client.
+  @Evolving
+  def checkCmReceiverPluginConfigured()  = {
+    if (ClientMetricsReceiverPlugin.getCmReceiver().isEmpty) {
+      throw new ClientMetricsReceiverPluginNotFoundException("Broker does not have any configured client metrics receiver plugin")
+    }
+  }
+
   def processGetTelemetrySubscriptionRequest(request: RequestChannel.Request,
                                              throttleMs: Int): GetTelemetrySubscriptionResponse = {
+    checkCmReceiverPluginConfigured()
     val subscriptionRequest = request.body[GetTelemetrySubscriptionRequest]
     val clientInfo = CmClientInformation(request, subscriptionRequest.getClientInstanceId.toString)
     _instance.processGetSubscriptionRequest(subscriptionRequest, clientInfo, throttleMs)
   }
 
+  def processPushTelemetryRequest(request: RequestChannel.Request, throttleMs: Int): PushTelemetryResponse = {
+    checkCmReceiverPluginConfigured()
+    val pushTelemetryRequest = request.body[PushTelemetryRequest]
+    val clientInfo = CmClientInformation(request, pushTelemetryRequest.getClientInstanceId.toString)
+    _instance.processPushTelemetryRequest(pushTelemetryRequest, request.context, clientInfo, throttleMs)
+  }
 }
 
 class ClientMetricsManager {
-  def getClientInstance(id: Uuid) = ClientMetricsCache.getInstance.get(id)
+
+  class ClientMetricsException(val s: String, var errorCode: Errors) extends KafkaException(s) {
+    def getErrorCode: Errors = this.errorCode
+  }
+
+  def getClientInstance(id: Uuid) : Option[CmClientInstanceState] =  {
+    Option(id) match {
+      case None | Some(Uuid.ZERO_UUID) => None
+      case _ => ClientMetricsCache.getInstance.get(id)
+    }
+  }
+
+  // Generates a new random client id and makes sure that it was not assigned to any of the previous clients
+  def generateNewClientId(): Uuid = {
+    var id = Uuid.randomUuid()
+    while (!ClientMetricsCache.getInstance.get(id).isEmpty) {
```

Comment:
The UUIDs should be so random as to make this check unnecessary.

But if it _is_ a necessary check, that implies (to me, anyway), that UUID collisions are common enough to warrant such a check. Which then seems to imply that we have a race condition where two incoming clients could be assigned the same UUID at roughly the same time, and then both later insert their data into the caches and the last one in wins.

Reply:
I agree, but to be on safer side I just check if it is already assigned in the cached objects.. but it is just an improvement. It sill does not cover the case that you have described or to make things worse, broker may not have any clientId in the cache as they would get cleaned up but client still carry the clientInstanceId in their requests ..